src/replication.c

```diff
@@ -2386,6 +2481,353 @@ char *sendCommandArgv(connection *conn, int argc, char **argv, size_t *argv_lens
     return NULL;
 }
 
+sds getReplicaPortString(void) {
+    long long replica_port;
+    if (server.slave_announce_port) {
+        replica_port = server.slave_announce_port;
+    } else if (server.tls_replication && server.tls_port) {
+        replica_port = server.tls_port;
+    } else {
+        replica_port = server.port;
+    }
+    return sdsfromlonglong(replica_port);
+}
+
+void freeReplDataBuf(void) {
+    if (server.pending_repl_data.blocks) {
+        listRelease(server.pending_repl_data.blocks);
+    }
+    server.pending_repl_data.blocks = NULL;
+    server.pending_repl_data.len = 0;
+}
+
+void abortRdbConnectionSync(int should_retry) {
+    if (server.repl_transfer_s) {
+        connClose(server.repl_transfer_s);
+        server.repl_transfer_s = NULL;
+    }
+    if (server.repl_full_sync_s) {
+        connClose(server.repl_full_sync_s);
+        server.repl_full_sync_s = NULL;
+    }
+    if (server.psync_master) {
+        freeClient(server.psync_master);
+        server.psync_master = NULL;
+    }
+    if (server.repl_transfer_fd != -1) {
+        close(server.repl_transfer_fd);
+        bg_unlink(server.repl_transfer_tmpfile);
+        zfree(server.repl_transfer_tmpfile);
+        server.repl_transfer_tmpfile = NULL;
+        server.repl_transfer_fd = -1;
+    }
+    server.repl_transfer_tmpfile = NULL;
+    server.repl_transfer_fd = -1;
+    freeReplDataBuf();
+
+    if (!should_retry) {
+        /* Make sure next time we will try different approach */
+        server.primary_can_sync_using_rdb_channel = 0;
+    }
+    return;
+}
+
+int reInitReplicaMainConnection(connection *conn) {
+    serverAssert(conn == server.repl_transfer_s);
+    /* Init the main connection for psync */
+    conn->state = CONN_STATE_CONNECTED;
+    server.repl_state = REPL_SEC_CONN_SEND_PSYNC;
+    serverAssert(connSetReadHandler(server.psync_master->conn, syncWithMaster) != C_ERR);
+    syncWithMaster(conn);
+    return C_OK;
+}
+
+int prepareRdbConnectionForRdbLoad(connection *conn) {
+    serverAssert(isRdbConnectionSync(conn));
+    /* Set connection handler for RDB */
+    serverAssert(connSetReadHandler(conn, readSyncBulkPayload) != C_ERR);
+    server.repl_transfer_size = -1;
+    server.repl_transfer_read = 0;
+    server.repl_transfer_last_fsync_off = 0;
+    server.repl_transfer_lastio = server.unixtime;
+    return C_OK;
+}
+
+int processEndOffsetResponse(char* response) {
+    long long reploffset;
+    char master_replid[CONFIG_RUN_ID_SIZE+1];
+    int dbid;
+    char endoff_format[30];
+    snprintf(endoff_format, sizeof(endoff_format), "$ENDOFF:%%lld %%%ds %%d", CONFIG_RUN_ID_SIZE);
+    if (sscanf(response, endoff_format, &reploffset, master_replid, &dbid) != 3) {
+        return C_ERR;
+    }
+    server.master_initial_offset = reploffset;
```

Comment:
In my opinion, there are two options for addressing this issue. We can either link the two connections or use a weaker approach, which will only link them to a specific offset. 
Using the first approach, master will generate a random magic string when bgsave begins and give it to the replica along with the `ENDOFFSET` response. It will keep a pointer to that end offset until the replica's main connection requests psync (using that magic string). 
The second approach involves master holding a struct containing offset and count. As soon as bgsave starts, master will insert the end offset into the struct and increment the count. This offset counter will be decremented when some replica requests psync with that offset.

The second approach might fail if some random replica asks for the exact offset, so I think the first approach is a bit more complex but generally better.

Reply:
i'm not sure i have the right context in my head right now, but if i get it correctly, with the second approach these are more issues with misbehaving clients. i feel i rather link them by communicating the client id, but maybe ip and port pair can work too.
please try to solve it (starting to collect slave buffers right away instead of relying on the backlog), and when the time comes i'll review the solution you chose and comment if necessary.

Reply:
btw, maybe we can create the two `client` structs together, one of them without a `conn` and reserve it for later (somewhat similar to what we do with cached_master).
not sure which one is cleaner.

Reply:
Ack, so I will continue with the first approach - peering replica's connection according to address & port tuples.

In the context of establishing two clients simultaneously, my understanding is that there might be a scenario where the master node ends up with three clients associated with a single replica. This arises because, during the initiation of the RDB connection by a replica, the master can only confirm that it is the same instance once the replica transmits its address. Consequently, upon accepting the new connection, the master may generate an additional, unused client before being able to identify and manage the connection correctly.